### PR TITLE
Adding support for symlinks and hardlinks

### PR DESCRIPTION
**Pull request by pombredanne**
There is a small bug in the current islnk: it return True for symlinks. The tarfile semantics are to return True for hardlinks and have issym returning True for symlinks

I also added support for linkpath and some tests.

### CODE REVIEW

**Comment by pombredanne** [comment]
@Changaco All ready for your review

**Comment by Changaco** [comment]
Please remove the explicit marking of unicode strings.

File: tests/test_entry.py [hunk 1]
```diff
@@ -5,9 +5,12 @@
 
 from __future__ import division, print_function, unicode_literals
 
+from os import path
 from os import stat
+from contextlib import closing
+import tarfile
 
-from libarchive import memory_reader, memory_writer
+from libarchive import memory_reader, memory_writer, file_reader
```
Comment (Changaco) [inline tests/test_entry.py]: The two `os` imports should be only one, and everything should be kept in alphabetical order.

File: libarchive/entry.py [hunk 2]
```diff
@@ -61,17 +61,36 @@ def isfifo(self):
 
     @property
     def islnk(self):
+        return True if ffi.entry_hardlink_w(self._entry_p) else False
```
Comment (Changaco) [inline libarchive/entry.py]: `True if something else False` is `bool(something)`

**Comment by pombreda** [comment]
@Changaco all set!

File: libarchive/entry.py [hunk 3]
```diff
@@ -61,17 +61,36 @@ def isfifo(self):
 
     @property
     def islnk(self):
+        return bool(ffi.entry_hardlink_w(self._entry_p))
+
+    @property
+    def issym(self):
         return self.filetype & 0o170000 == 0o120000
```
Comment (Changaco) [inline libarchive/entry.py]: Shouldn't this be `bool(ffi.entry_symlink_w(self._entry_p))`?

File: libarchive/entry.py [hunk 4]
```diff
@@ -61,17 +61,36 @@ def isfifo(self):
...
 
     @property
+    def linkpath(self):
+        if self.islnk:
+            return ffi.entry_hardlink_w(self._entry_p)
+        if self.issym:
+            return ffi.entry_symlink_w(self._entry_p)
```
Comment (Changaco) [inline libarchive/entry.py]: `return ffi.entry_hardlink_w(self._entry_p) or ffi.entry_symlink_w(self._entry_p)` would be more efficient.

Also, what's the use case for `linkpath`? Why not have two separate properties instead?

Comment (pombreda) [inline libarchive/entry.py]: @Changaco The use case for linkpath is to have the same api as tarfile that uses a single linkpath for either hard of sym links, except that tarfile does not use properties by methods, at least keep the same names.

File: libarchive/entry.py (hunk 3 above)
Comment (pombreda) [inline libarchive/entry.py]: @Changaco nope, at least not if you want to keep the attribute names the same as in tarfile: 
islnk is for hardlinks and isym for symlinks

Comment (pombreda) [inline libarchive/entry.py]: You can see the tests I added for this. FWIW this was the issue in your initial islnk implementation

File: libarchive/entry.py (hunk 4 above)
Comment (pombreda) [inline libarchive/entry.py]: @Changaco About efficiency I am not sure whether it is more efficient or not.
The definition of the symlink is as a file type whereas the hardlink is through a test for the path existence.  Not sure which one would be fasted. In anycase the test for symlink would have to be on type first so I find this way explicit enough without possible gotchas.

File: libarchive/entry.py (hunk 3 above)
Comment (Changaco) [inline libarchive/entry.py]: What is the value of `filetype` for hard links? Is it the type of the target file?

File: libarchive/entry.py (hunk 4 above)
Comment (Changaco) [inline libarchive/entry.py]: What I'm seeing in the `tarfile` module is `linkname`, not `linkpath`.

Comment (Changaco) [inline libarchive/entry.py]: > About efficiency I am not sure whether it is more efficient or not.

I'm sure, the current code calls `ffi.entry_hardlink_w` twice in the case of a hard link.

Comment (pombreda) [inline libarchive/entry.py]: entirely True.... I guess I am too used to have properties cached at times.

File: libarchive/entry.py (hunk 3 above)
Comment (pombreda) [inline libarchive/entry.py]: Typically null, but it is undefined. At the time you read some entry the linked-to-file is by construction something not yet known until you read the corresponding file. See https://github.com/libarchive/libarchive/wiki/Hardlinks for details

File: libarchive/entry.py (hunk 4 above)
Comment (pombreda) [inline libarchive/entry.py]: About linkpath: https://hg.python.org/cpython/file/b9b772a91db0/Lib/tarfile.py#l769

**Comment by pombreda** [comment]
just realized that you tests against a really old libarchive12 ?

**Comment by Changaco** [comment]
I test against what's available in the Travis environment. It took me a while to figure it out, I don't use Ubuntu on my machines.

**Comment by pombredanne** [comment]
@Changaco so the failure does not seem to be related to the libarchive version, but rather your Unicode imho.
an archive entry path can be arbitrary binary string and the encoding is unknown in most cases since it was eventually the local FS encoding used by whoever created the archive. Furthermore, each entry path can have a different encoding.

Stuffing that in Python unicode is unlikely to have a happy outcome, which is IMHO the reason why the tests added fails and that some entry paths are None. 

Natively libarchive handles these happily. I think the problem here is only rely on unicode _w calls may be?

**Comment by pombredanne** [comment]
Just to add to this: you can never ever be able to fit *all* the arbitrary binary paths that exists out there in archives in a valid Unicode string afaik becasue a Linux FS path is an arbitrary binary string that can be anything and archive entries can be anything too.

**Comment by pombredanne** [comment]
@Changaco btw, tell me if I am being annoying or too intense with all these bugs and PRs... If that is so I can just quietly maintain and evolve my fork.

**Comment by pombredanne** [comment]
I am going to create a new PR that deals only with hard/symlinks

**Comment by pombredanne** [comment]
@Changaco fixing the handling of hard/symlinks and the handling of non-unicode paths is really hard to separate as paths handlings happens in the same places.... Do you mind if I create a clean PR that handles both?

**Comment by Changaco** [comment]
I don't see why you'd need a new PR, you can rebase this one if necessary, but by all means do what you think is best.

**Comment by pombredanne** [comment]
So based on this https://travis-ci.org/Changaco/python-libarchive-c/jobs/62066990 it sounds clear to me that it will be rather hard to use a single codebase for 2.7 and 3.x ... @Changaco are you attached to keeping support for Python 3.x?

**Comment by pombredanne** [comment]
This failure https://travis-ci.org/Changaco/python-libarchive-c/jobs/62066989#L608 may be fixable and is likely test-related.

**Comment by pombredanne** [comment]
This failure https://travis-ci.org/Changaco/python-libarchive-c/jobs/62066989#L761 may be an updtream bug of sorts.

**Comment by pombredanne** [comment]
This other failure is likely an upstream bug too:
https://travis-ci.org/Changaco/python-libarchive-c/jobs/62066989#L798

**Comment by pombredanne** [comment]
@Changaco about the flake8 failure, I do not think that pep8'ing test code for line length make sense in general:
https://travis-ci.org/Changaco/python-libarchive-c/jobs/62066989#L851

So understand why this is not always a good thing, check out: http://www.youtube.com/watch?v=wf-BqAjZb8M

**Comment by pombredanne** [comment]
I filed this bug with libarchive: https://github.com/libarchive/libarchive/issues/538

**Comment by Changaco** [comment]
I've written quite a few pieces of code that work on both versions of python, there's always a solution, and it's usually quite simple.

**Comment by Changaco** [comment]
> about the flake8 failure, I do not think that pep8'ing test code for line length make sense in general

I'll accept a PR that turns off the line length check.

File: libarchive/entry.py [hunk 5]
```diff
@@ -61,30 +61,53 @@ def isfifo(self):
...
+
+    @property
     def mtime(self):
         return ffi.entry_mtime(self._entry_p)
 
-    @property
-    def pathname(self):
-        return ffi.entry_pathname_w(self._entry_p)
+    def _getpathname(self):
+        return ffi.entry_pathname(self._entry_p)
```
Comment (Changaco) [inline libarchive/entry.py]: I think what we want is `return ffi.entry_pathname_w(self._entry_p) or ffi.entry_pathname(self._entry_p)`.

File: tests/test_entry.py [hunk 6]
```diff
@@ -24,6 +31,162 @@ def test_entry_properties():
...
+
+def decode_path(arch_path):
+    """
+    Return a `arch_path` bytes string decoded from a base64-encoded unicode string.
+    Rationale: tests expectations are stored as UTF-8 JSON yet paths can be
+    arbitrary byte strings. This encoding ensures that we can safely store
+    bytes in UTF-8 strings.
+    """
+    if arch_path:
+        arch_path = bytes(codecs.decode(arch_path, 'base64'))
```
Comment (Changaco) [inline tests/test_entry.py]: You shouldn't call the `bytes` function, it doesn't work the same way in both python versions.

File: tests/test_entry.py [hunk 7]
```diff
@@ -5,9 +5,16 @@
 
 from __future__ import division, print_function, unicode_literals
 
-from os import stat
+import codecs
```
Comment (Changaco) [inline tests/test_entry.py]: `from base64 import b64decode, b64encode` would be better.

**Comment by pombredanne** [comment]
about the https://github.com/libarchive/libarchive/issues/538 and the fact that the test on unicode.zip return no pathname whether using pathname or pathname_w on Linux, I think this is a bug somewhere in Python. Why?  python-libarchive that use SWIG exhibits the same exact issue for this archive. bsdtar which is pure C and does not marshall to Python does not exhibit the problem

Note that bsdtar never uses the pathname_w function, so using it is not an option IMHO. And in practice it is most often empty anyway on Linux.

That said see also https://github.com/libarchive/libarchive/issues/538#issuecomment-101184878

File: libarchive/entry.py (hunk 5 above)
Comment (pombredanne) [inline libarchive/entry.py]: IMHO no . I do not think we ever want the _w version. In practice the plain version is always populated. All code in bsdtar and bsdcpio never use the wide version.

File: tests/test_entry.py (hunk 6 above)
Comment (pombredanne) [inline tests/test_entry.py]: Well that is rather painful ...  I will make things work on Py2 first

File: tests/test_entry.py (hunk 7 above)
Comment (pombredanne) [inline tests/test_entry.py]: well they are essentially the same, but I am fine with it

**Comment by pombredanne** [comment]
About 3.4 failures, I have no idea how to fix them. Can you look into these?
https://travis-ci.org/Changaco/python-libarchive-c/jobs/62251185

About this failure here https://travis-ci.org/Changaco/python-libarchive-c/jobs/62251183#L197
This is passing with the latest libarchive, so to me this is not a bug.

**Comment by pombredanne** [comment]
@Changaco at this stage I am satisfied with the this working on 2.7 and the latest stable libarchive 3.1.2
I am leaving it to you if you want to support 3.4 and other libarchive versions.

**Comment by Changaco** [comment]
Closing in favor of #20.

**Pull request closed without merging**